examples/setup_run.py

```diff
@@ -26,6 +26,61 @@
                            Optional[Union[ReplayBuffer, PrioritisedReplayBuffer]]], Plotter]
 
 
+def get_n_iterations(
```

Comment:
A function that gets the number of training iterations, with an option to calculate this based of the number of flow forward passes required.

Reply:
Don't we need more log prob evaluations when using HMC? For each iteration, we need `n_transition_operator_inner_steps + 1` for the leapfrog and then an additional one to do decide about acceptance. For the latter, we also need an additional step in the beginning, so it should be `(n_transition_operator_inner_steps + 1) * n_intermediate_ais_dist + 1` evaluations, right? And what about sampling?

Reply:
Yeah - good spot!

Reply:
So for HMC it should be

Reply:
Yeah so for both hmc and metropolis we need an additional forward pass for the initial sampling from the flow (via `sample_and_log_prob` - I've updated the code to fix this. But for HMC we do not need `n_transition_operator_inner_steps + 1` for the following reason: For each x point we could call the equivalent of jax's `value_and_grad` method when computing the forward pass of the flow (either for sampling to init the AIS chain, or for log prob evaluation), and then pass on the value/grad to any future parts of the code that either need grad (e.g. HMC inner step) or log prob (e.g. accept/reject criterion). Thus we just need `(n_transition_operator_inner_steps) * n_intermediate_ais_dist + 1` flow forward passes.

Reply:
I've also accounted for the batch size now

Reply:
![image](https://user-images.githubusercontent.com/44415540/173184601-1a1d40dc-bc3e-4de3-bdab-13c3ac9cd93e.png)

Reply:
The point here is that per unique x that we generate during AIS (including part of the inner HMC step) in theory we only need one forward pass of the flow. In practice we re-do the flow forward pass many more times than what is necessary (e.g. when calculating AIS weight we typically re-evaluate the flow rather than taking the log prob that we used already in the metropolis accept/reject step).

Reply:
Ah, I see, you're right. Thanks for having another close look on this, @lollcat!